libsnark/zk_proof_systems/plonk/srs.tcc

```diff
@@ -56,38 +54,194 @@ srs<ppT>::srs(
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+template<typename ppT>
+void transcript_hasher<ppT>::add_element(const libff::G1<ppT> &element)
+{
+    libff::G1<ppT> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
 template<typename ppT>
-transcript_hash_t<ppT>::transcript_hash_t(
-    libff::Fr<ppT> &beta,
-    libff::Fr<ppT> &gamma,
-    libff::Fr<ppT> &alpha,
-    libff::Fr<ppT> &zeta,
-    libff::Fr<ppT> &nu,
-    libff::Fr<ppT> &u)
-    : beta(beta), gamma(gamma), alpha(alpha), zeta(zeta), nu(nu), u(u)
+void transcript_hasher<ppT>::add_element(const libff::G2<ppT> &element)
+{
+    libff::G2<ppT> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+template<typename ppT> libff::Fr<ppT> transcript_hasher<ppT>::get_hash()
 {
+    size_t buffer_len = this->buffer.size();
+    // DEBUG
+    printf("[%s:%d] len %7d\n", __FILE__, __LINE__, (int)buffer_len);
+
+    // vector of valid lengths (\attention specific to BLS12-381)
+    const std::vector<size_t> length{288, 320, 416, 704, 896, 1120};
+
+    // If we are here, then the hasher buffer has invalid length so throw an
+    // exception
+    bool b_valid_length =
+        (0 != std::count(length.begin(), length.end(), buffer_len));
+    if (!b_valid_length) {
+        throw std::logic_error(
+            "Error: invalid length of transcript hasher buffer");
+    }
+    if (!b_valid_length) {
+        printf(
+            "[%s:%d] Error: invalid length of transcript hasher buffer\n",
+            __FILE__,
+            __LINE__);
+    }
+    assert(b_valid_length);
+
+    libff::Fr<ppT> challenge = 0;
+
+    // beta
+    if (buffer_len == length[0]) {
+        printf(
+            "[%s:%d] buffer_len %d: beta\n",
+            __FILE__,
+            __LINE__,
+            (int)buffer_len);
+        challenge = this->hash_values[0]; // beta
+    }
+    // gamma
+    if (buffer_len == length[1]) {
```

Comment:
The only reason to repeat all this code for each case seems to be for the variable names . If you want to keep the debug output, how about we create a vector of names.

Reply:
Addressed using a `map<size_t, string>` and a `while` loop.